lib/setup.c

```diff
@@ -921,6 +926,40 @@ ___SIZE_TS requested_words_still;)
 }
 
 
+___EXP_FUNC(___BOOL, ___fdset_resize)
+   ___P((int fd1,
+         int fd2),
+        (fd1,
+         fd2)
+int fd1;
+int fd2;)
+{
+#ifdef USE_select_or_poll
+
+  ___processor_state ___ps = ___PSTATE;
+  int fd = (fd2 > fd1) ? fd2 : fd1;
+
+  if (fd < ___ps->os.fdset.size)
+    return 0;
+
+  ___fdset_resize_heap_overflow_clear ();
+
+  ___sync_op_struct sop;
+
+  sop.op = OP_FDSET_RESIZE;
+  sop.arg[0] = fd;
+
+  on_all_processors (___PSP &sop);
+
+  return ___fdset_resize_heap_overflow ();
```

Comment:
It's possible for two threads to enter `___fdset_resize` simultaneously, before being serialized by `on_all_processors`, which is the same as the race you describe with two of them happening in quick succession.

Basically what I propose is something like this in pseudo, where `__fdset_resize_start` needs to be atomic inc and `__fdset_resize_end` needs to be atomic decrement :
```
 if (!__fdset_resizing_start ())
   __fdset_clear_overflow ()

do __fdset_resize ...

overflow = ___fdset_overflow()
if (___fdset_resize_end ())
  ___fdset_clear_overflow()

return overflow
```